torch/_tensor_docs.py

```diff
@@ -2651,11 +2651,26 @@ def callable(a, b) -> number
 
 add_docstr_all('numpy',
                r"""
-numpy() -> numpy.ndarray
+numpy(*, force=False) -> numpy.ndarray
 
-Returns :attr:`self` tensor as a NumPy :class:`ndarray`. This tensor and the
+Returns :attr:`self` tensor as a NumPy :class:`ndarray`.
+
+If :attr:`force` is ``False``, tensor :attr:`self` and the
 returned :class:`ndarray` share the same underlying storage. Changes to
 :attr:`self` tensor will be reflected in the :class:`ndarray` and vice versa.
+If tensors storage is not CPU, or if :attr:`requires_grad` is ``True`` and
```

Comment:
"If tensors storage" -> "If the tensor's storage"

Reply:
> The tensor and the returned ndarray will only share their storage if the conversion would have happened with :attr:`force` being False.

I don't think that's true. If we have tensor that is:

- dense
- on CPU
- not conjugated
- requires grad

then t.numpy() throws an exception, but t.numpy(force = True) returns ndarray that shares its storage.

Reply:
".. if the conversion would have happened with :attr:`force` being False and the tensor being detached" ?

Reply:
In the last commit i went with:

If the tensor
isn't on the CPU or its conjugate bit is set, the tensor won't share
its storage with the returned ndarray.

Reply:
I think it's an interesting UX question whether a tensor with requires_grad=True should share its storage with the NumPy array created or not. I kinda lean towards creating a copy in that case for clarity and to avoid troublesome scenarios later.

What do you think, @albanD?

Reply:
In the case of `requires_grad=True` (and if other view conditions are met), I suggest returning a read-only numpy array (set [writable flag](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.setflags.html#numpy-ndarray-setflags) to False) that shares storage with the tensor in order to avoid copies.

Reply:
If I look back to the issues and all previous communication, I think, that the purposes of the force argument are

- to be a shorthand for `tensor.detach().resolve_conj().cpu().to_dense().numpy()`
- to give make it easier to work with many libraries at once, so if you are trying to create some visualization, you don't care if you  have a Torch tensor, or a TensorFlow tensor or a Panda array or whatever ... you just call numpy(force=True)

We have already "broken" that goal by letting sparse tensors throw an exception insight of potential future BC problems. I don't think we should break it again if we don't have a serious reason. And I don't see any reason.

Reply:
I think this is ok to have a regular numpy array that shares data.
It is just like when you do `.detach()` in general, any inplace change is not differentiable and won't be tracked. You get the same behavior here.

Reply:
And I don't think we want to diverge from the `tensor.detach().resolve_conj().cpu().numpy()` definition.

Reply:
For what it's worth, `to_dense()` is not always realizable, for instance, when a sparse tensor has many large dimensions and one may not have enough RAM to store such a tensor using a strided layout.

Different from a conjugate bit, a sparse layout is not a lazy variant of a dense tensor. The result of `to_dense()` operation may depend on how are the unspecified elements interpreted. For instance, are they zeros or do sparse tensors represent masked tensors? In the latter case, one may want to get a masked numpy array as a result of `.numpy()`. If `numpy(force=True)` will involve `to_dense`, one has to fix the semantics of sparse tensors. I am not sure that this PR, that main objective is to introduce a shorthand to a convenience transformation, is the proper place to make these decisions.

Reply:
Notice that I didn't put `to_dense()` in the snippet above ;) 
I agree with you that it is a much trickier business and we can leave that to a later PR.

Reply:
Yes, @albanD , I noticed it but only after pressing the Comment button ;)